node/src/api/registration.rs

```diff
@@ -1,181 +1,118 @@
+        let request = resource.registration_request(api_version);
+
+        Ok(client
+            .post(url.clone())
+            .json(&request)
+            .send()
+            .await?
+            .error_for_status()?)
     }
 
-    async fn register_sender(
+    pub async fn delete_resource(
         client: &reqwest::Client,
         url: &reqwest::Url,
-        sender: &resource::Sender,
-    ) -> Result<(), Box<dyn std::error::Error>> {
-        use nmos_model::version::is_04::V1_0;
-        use nmos_schema::is_04::v1_0_x::{
-            RegistrationapiResourcePostRequest, RegistrationapiResourcePostRequestHealthVariant2,
-        };
-
-        let resource::SenderJson::V1_0(sender_json) = sender.to_json(&V1_0);
-        let sender_post_request = RegistrationapiResourcePostRequestHealthVariant2 {
-            data: Some(sender_json),
-            type_: Some(String::from("sender")),
-        };
-        let post_request = RegistrationapiResourcePostRequest::Variant2(sender_post_request);
-
-        client.post(url.clone()).json(&post_request).send().await?;
-
-        Ok(())
+        resource: &dyn resource::Registerable,
+    ) -> Result<reqwest::Response, Box<dyn std::error::Error>> {
+        let delete_url = url
+            .clone()
+            .join(format!("resource/{}", resource.registry_path()).as_str())
+            .unwrap();
+
+        Ok(client.delete(delete_url).send().await?.error_for_status()?)
     }
 
-    async fn register_receiver(
+    pub async fn register_resource(
         client: &reqwest::Client,
         url: &reqwest::Url,
-        receiver: &resource::Receiver,
-    ) -> Result<(), Box<dyn std::error::Error>> {
-        use nmos_model::version::is_04::V1_0;
-        use nmos_schema::is_04::v1_0_x::{
-            RegistrationapiResourcePostRequest, RegistrationapiResourcePostRequestHealthVariant3,
-        };
-
-        let resource::ReceiverJson::V1_0(receiver_json) = receiver.to_json(&V1_0);
-        let receiver_post_request = RegistrationapiResourcePostRequestHealthVariant3 {
-            data: Some(receiver_json),
-            type_: Some(String::from("receiver")),
-        };
-        let post_request = RegistrationapiResourcePostRequest::Variant3(receiver_post_request);
-
-        client.post(url.clone()).json(&post_request).send().await?;
+        resource: &dyn resource::Registerable,
+        api_version: &APIVersion,
+    ) -> Result<reqwest::Response, Box<dyn std::error::Error>> {
+        let res = Self::post_resource(client, url, resource, api_version).await?;
+
+        if res.status() == StatusCode::OK {
+            warn!(
+                "Resource already present in API deleting: {}",
+                resource.registry_path()
+            );
+
+            let res = Self::delete_resource(client, url, resource).await?;
+
+            if res.status() == StatusCode::OK {
+                return Err(Box::new(std::io::Error::new(
```

Comment:
I do like the generic Error approach here but I would not use std::io::Error because it does not really match the semantics of this Error. You could either provide a custom error type or create one with the help of dedicated error crates. Since
we do not really differentiate Errors here I personally would use anyhow.

The nice thing on anyhow is that they also implement std::error::Error and are also usable in threaded contexts.

Reply:
Yea i was thinking about switching to anyhow errors everywhere but opted to skip it for "effort"-reasons